Sources/Core/EthereumABI/ABIEncoding.swift

```diff
@@ -186,6 +191,29 @@ extension ABIEncoder {
         return headsConcatenated + tailsConcatenated
     }
 
+    /// Performs ABI encoding conforming to [the documentation of encoding](https://docs.soliditylang.org/en/develop/abi-spec.html#basic-design) in Solidity
+    /// but **it does not add the data offset for dynamic types!!** To return single value **with data offset** use the following instead:
+    /// ```
+    /// ABIEncoder.encode(types: [type], values: [value] as [AnyObject])
+    /// ```
+    /// Almost identical to use of `web3.eth.abi.encodeParameter` in web3.js.
+    /// Calling `web3.eth.abi.encodeParameter('string','test')` in web3.js will return the following:
+    /// ```
+    /// 0x0000000000000000000000000000000000000000000000000000000000000020
+    /// 0000000000000000000000000000000000000000000000000000000000000004
+    /// 7465737400000000000000000000000000000000000000000000000000000000
+    /// ```
+    /// but calling `ABIEncoder.encodeSingleType(type: .string, value: "test" as AnyObject)` will return:
+    /// ```
+    /// 0x0000000000000000000000000000000000000000000000000000000000000004
+    /// 7465737400000000000000000000000000000000000000000000000000000000
+    /// ```
+    /// - Parameters:
+    ///   - type: Solidity type of the `value`;
+    ///   - value: value to encode.
+    /// - Returns: ABI encoded data, e.g. function call parameters. Returns `nil` if:
+    ///     - `types.count != values.count`;
+    ///     - encoding has failed (e.g. type mismatch).
```

Comment:
This is basically what triggered me to write docs and make sure people will know how each function works.
Don't recall exactly what I was implementing but I expected that `encodeSingleType` would encode dynamic types with data offset in front of the data itself but it didn't happen. Was wondering why? Is it a bug? Appears that it's just different from web3.js.

While writing the doc I also came up with another idea that will allow us to get the ABI with data offset from this function. The idea is the following: add a new input argument `dynamicDataOffset` that will be `true` by default
```
encodeSingleType(type: ABI.Element.ParameterType, value: AnyObject, dynamicDataOffset: Bool = true)
```

Uses of this function across the library will have to be updated to with `encodeSingleType(..., dynamicDataOffset: false)`.

On the other hand, this update is a breaking change that will be silently pushed and could potentially break other projects. So as a win-win situation I think it would make sense to push that idea with the new argument to the v4.